### PR TITLE
ext2: Fix _ext2_dir_read() skipping last entries in a directory

### PR DESCRIPTION
**Pull request by astalke**
DONE: RTOS-322

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: ext2/dir.c [hunk 1]
```diff
@@ -129,11 +129,16 @@ int _ext2_dir_read(ext2_t *fs, ext2_obj_t *dir, offs_t offs, struct dirent *res,
 	if ((entry = (ext2_dirent_t *)malloc(len)) == NULL)
 		return -ENOMEM;
 
-	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) != len) {
+	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < sizeof(ext2_dirent_t)) {
```
Comment (nalajcie) [inline ext2/dir.c:132]: nice error due to implicit integer promotion :).
- assume `ret = -1` 
-  You're comparing against `size_t` (unsigned type), `ret` will be promoted to unsigned type - UINT_MAX value
- the actual comparison (`UINT_MAX < sizeof(ext_dirent_t)`) would be **false**

See https://wiki.sei.cmu.edu/confluence/display/c/INT02-C.+Understand+integer+conversion+rules for more info and how it should be fixed.

(note: previous code was also invalid, but due to using `!=` would not fail so spectacularly :))

Comment (astalke) [reply ext2/dir.c]: I've made a fix. Can you check if it is correct now?

File: ext2/dir.c [hunk 2]
```diff
@@ -129,9 +129,14 @@ int _ext2_dir_read(ext2_t *fs, ext2_obj_t *dir, offs_t offs, struct dirent *res,
...
 		return -ENOMEM;
 
-	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) != len) {
+	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < 0) {
 		free(entry);
-		return (ret < 0) ? (int)ret : -ENOENT;
+		return ret;
+	}
+
+	if (((size_t)ret < sizeof(ext2_dirent_t)) || ((size_t)ret < sizeof(ext2_dirent_t) + entry->len)) {
```
Comment (agkaminski) [inline ext2/dir.c:137]: `((size_t)ret < sizeof(ext2_dirent_t))` condition is redundant (if `entry->len` >= 0)
```suggestion
	if ((size_t)ret < sizeof(ext2_dirent_t) + entry->len) {
```

Comment (nalajcie) [reply ext2/dir.c:137]: not really - if the first part is true You won't access garbage value `entry->len`.

Still - not readable enough (the check might be deleted during next refactoring) - eg. by @agkaminski  :P

IMHO would be better to:
```c
if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < (ssize_t)sizeof(ext2_dirent_t)) {
  // handle read error / partial read
}
// other checks accessing `entry` follow
```

Side note - to be MISRA-compliant the assignment should be done before the `if`

Comment (astalke) [reply ext2/dir.c]: > not really - if the first part is true You won't access garbage value `entry->len`.
> 
> Still - not readable enough (the check might be deleted during next refactoring) - eg. by @agkaminski :P
> 
> IMHO would be better to:
> 
> ```c
> if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < (ssize_t)sizeof(ext2_dirent_t)) {
>   // handle read error / partial read
> }
> // other checks accessing `entry` follow
> ```
> 
> Side note - to be MISRA-compliant the assignment should be done before the `if`

I've just moved the assignment before the `if` and split the condition.

**Review by agkaminski — APPROVED** [review]
LGTM, let's wait for @Maxez review

**Comment by agkaminski** [comment]
Fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/581

**Review by Maxez — APPROVED** [review]
Good catch with the signed/unsigned comparison. LGTM too.

**Pull request merged**